src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java

```diff
@@ -624,25 +625,32 @@ private Future<Boolean> createHoldingsForChangedLocations(Piece piece, String in
       return Future.succeededFuture(true);
     }
 
-    String receivingTenantId = getReceivingTenantId(piece);
-    String locationId = getLocationId(piece);
-    String holdingId = getHoldingId(piece);
-    logger.info("createHoldingsForChangedLocations:: receivingTenantId: {} locationId: {} holdingId: {}",
-      receivingTenantId, locationId, holdingId);
-    Location location = new Location().withLocationId(locationId).withHoldingId(holdingId);
-    var locationContext = RequestContextUtil.createContextWithNewTenantId(requestContext, receivingTenantId);
+    String srcTenantId = piece.getReceivingTenantId();
+    String dstTenantId = getReceivingTenantId(piece);
+    Location location = new Location().withLocationId(getLocationId(piece)).withHoldingId(getHoldingId(piece));
+    RequestContext locationContext = RequestContextUtil.createContextWithNewTenantId(requestContext, dstTenantId);
+    logger.info("createHoldingsForChangedLocations:: Creating shadow instance and holding if needed, srcTenantId: {}, dstTenantId: {}, locationId: {}, holdingId: {}",
+      srcTenantId, dstTenantId, location.getLocationId(), location.getHoldingId());
     return inventoryInstanceManager.createShadowInstanceIfNeeded(instanceId, locationContext)
-      .compose(instance -> inventoryHoldingManager.getOrCreateHoldingsRecord(instanceId, location, locationContext))
+      .compose(instance -> {
+        if (Objects.isNull(srcTenantId) || (srcTenantId.equals(dstTenantId))) {
+          return inventoryHoldingManager.getOrCreateHoldingsRecord(instanceId, location, locationContext);
```

Comment:
Doesn't create holding for all pieces

Reply:
comment can be added in source code also, because github comment can be forgotten